### DISCUSSION TITLE
Installing autotag

### DISCUSSION THREAD

**Discussion by r3k4ce** [LazyVim/LazyVim#644] [Q&A] [answered]
Hi guys,

Probably a noob question, but I am having trouble understanding how to install this plugin https://github.com/windwp/nvim-ts-autotag using the starter configuration for lazy.nvim. I have been able to install other plugins by following the documentation on the official site. I am essentially trying to get the same result I would get in VScode when inserting an html tag. For example when inserting a tag for html code in VScode it will automatically add the closing tag.

The documentation for the github page advises the following to get this working.

Neovim 0.7 and nvim-treesitter to work

User treesitter setup

require'nvim-treesitter.configs'.setup {
  autotag = {
    enable = true,
  }
}

I tried copying the tree sitter spec from the lazy.nvim https://www.lazyvim.org/plugins/treesitter site into a lua file inside the plugins folder of the lazy.nvim starter structure and then inserting the autotag = {enable = true} in the opts section of the full spec for tree sitter but this did not work.

**Comment by dpetka2001** [comment, marked as answer]
Put the following into a file in `~/.config/lua/plugins` and you should be good to go
```lua
return {
  -- Modify Treesitter
  {
    "nvim-treesitter/nvim-treesitter",
    dependencies = {
      -- Add nvim-ts-autotag
      { "windwp/nvim-ts-autotag" },
    },
    event = { "BufReadPre" },
    opts = function(_, opts)
      vim.list_extend(opts.ensure_installed, {
        "c",
        "cpp",
        "fish",
        "rust",
      })
      return vim.tbl_deep_extend("force", opts, {
        autotag = {
          enable = true,
        },
      })
    end,
  },
}
```

**Reply by r3k4ce** [reply]
This worked perfectly. How did you know how to go about writing this script? Did you already have this plugin installed?

**Reply by dpetka2001** [reply]
Yes I already have it installed and when I started I took inspiration by looking at other people's configurations. I still am not proficient at Lua but I know some few things to customize my configuration.

**Comment by Jomik** [comment]
Just going to provide two alternatives, that seem cleaner than the marked answer.
I would probably not do a full override of `nvim-treesitter`'s options, unless you know why you are doing it.
Also no reason to override the `event`, unless you intentionally want to remove an event here.
https://github.com/LazyVim/LazyVim/blob/86ac9989ea15b7a69bb2bdf719a9a809db5ce526/lua/lazyvim/plugins/treesitter.lua#L6

```lua
return {
"nvim-treesitter/nvim-treesitter",
    dependencies = {
      "windwp/nvim-ts-autotag",
    },
    opts = {
      autotag = {
        -- Setup autotag using treesitter config.
        enable = true,
      },
    },
}
```
or
```lua
return {
  {
    "nvim-treesitter/nvim-treesitter",
    dependencies = {
      "windwp/nvim-ts-autotag",
      -- Setup autotag with default options.
      opts = {},
    },
  }
}
```

**Reply by dpetka2001** [reply]
Other than the `event` option which I honestly don't remember why I overrode and you're definitely right we shouldn't override defaults unless we know what we're specifically doing, there's no other overriding with the rest of the config. It just extends treesitter's default configuration.

**Reply by Jomik** [reply]
Agreed, but your extension to `ensure_installed` has nothing to do with `autotag`. And if we do not add to a list, there is no reason to do an `opts` function, as lazy will automatically merge tables. 👍

**Reply by dpetka2001** [reply]
Agreed, I just pasted it like this as a reference in case OP wanted to do some other changes (for example to `ensure_installed` and addtional keys to `opts`), so that he doesn't have to make 2 different specs to achieve that. If all he cares about is the `autotag` functionality your solution is much cleaner.
